controllers/imagepolicy_controller.go

```diff
@@ -148,6 +148,20 @@ func (r *ImagePolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 
 	var err error
```

Comment:
Is this shadowed right away? If I'm reading this right, probably worth just removing this declaration as part of this change, for clarity. And using a differently-named var to assign to in the loop a few lines below.

Reply:
OK I think I see what's going on. The expectation is that the value returned from `PolicerFromSpec` _either_ has non-nil `policer` or non-nil `err`; so if there's an error, the block which accesses policer won't run, and the conditional following that will catch errors from parsing the policy _or_ from applying the policy. All else being correct, this would work fine.

However, `policer` is typed as `Policer`, whereas the return value is `*SemVer`; this type checks, but `nil(Policer) != nil(*SemVer)`, so the block conditional on `policer != nil` runs after all.  Am I right?

Reply:
> However, policer is typed as Policer, whereas the return value is *SemVer; this type checks, but nil(Policer) != nil(*SemVer), so the block conditional on policer != nil runs after all. Am I right?

Yes I think I am. If I remove the early exit you added, and put `fmt.Printf("%s\n", policer)` _inside_ the block conditional on `policer != nil`, I can see 

    [DEBUG] %!s(*policy.SemVer=<nil>)

printed in the output of

    go test -v ./controllers/...

(and a rightly failing test). Though the fix here already avoids it crashing, I think it'd be better to fix that underlying problem as well.

Reply:
Yeah, that's seems to be what was happening. And later on when `policer.Latest()` is called, it panicked. An interface holding a nil value not equal to nil. The policer nil check will only protect against the situation where the choice is something other than the known choices, but I don't see that'll really ever happen. We should be able to remove the policer check.

Reply:
I think it would pay to make `PolicerFromSpec` return Policer(nil) values when it gets an error, otherwise this will remain as a trap for future changes.

Reply:
Yeah, adding that. Thanks for digging into it.